### PR TITLE
Add initial support for the Azure instance metadata service.

### PR DESCRIPTION
**Pull request by brendandburns**
Part of fixing #46632

@colemickens @rootfs @jdumars @kris-nova

### CODE REVIEW

File: pkg/cloudprovider/providers/azure/azure_instances.go [hunk 1]
```diff
@@ -29,6 +29,16 @@ import (
 
 // NodeAddresses returns the addresses of the specified instance.
 func (az *Cloud) NodeAddresses(name types.NodeName) ([]v1.NodeAddress, error) {
+	if az.UseInstanceMetadata {
+		text, err := QueryMetadataText("instance/network/interface/0/ipv4/ipAddress/0/privateIpAddress")
```
Comment (liggitt) [inline pkg/cloudprovider/providers/azure/azure_instances.go]: if this errors, should it fall back to the previous API calls or return an error?

Comment (liggitt) [inline pkg/cloudprovider/providers/azure/azure_instances.go]: the "where possible" language in the config option is squishy, so I wasn't sure

Comment (brendandburns) [inline pkg/cloudprovider/providers/azure/azure_instances.go]: Where possible was meant to mean: "where the metadata service provides the information" as opposed to "fallback if there are errors"

We don't fallback in other places where we use other metadata services:

https://github.com/kubernetes/kubernetes/blob/master/pkg/cloudprovider/providers/aws/aws.go#L779

So I think we probably don't want to here, either.

**Comment by colemickens** [comment]
LGTM, but out of curiosity, why implement the various networking structs if the only usage of the IMDS is to directly grab the relevant ipv4 needed and you never actually use/deserialize the containing structs?

/lgtm

**Comment by brendandburns** [comment]
I'm going to pull more eventually...

(also eventually, I'll fork it out as it's own library)

File: pkg/cloudprovider/providers/azure/azure_instance_metadata.go [hunk 2]
```diff
@@ -0,0 +1,103 @@
+/*
+Copyright 2016 The Kubernetes Authors.
```
Comment (rootfs) [inline pkg/cloudprovider/providers/azure/azure_instance_metadata.go]: 2017

**Pull request merged**